JOSS paper (#71)
FAIRmat-NFDI/nyaml: paper/paper.md

---
title: "nyaml: Format Converter for the NeXus Data Model"
tags:
  - Python
  - NeXus
  - NXDL
  - YAML
  - XML

authors:
  - given-names: Rubel
    surname: Mozumder
    orcid: https://orcid.org/0009-0007-5926-6646
    affiliation: 1

>     equal-contrib: true

Reviewer: We cannot have 5 people with equal contribution! That is saying that five people are the lead author.

I think for this particular paper we have agreed that Rubel is the sole main author. So please remove equal-contrib everywhere

Author: Yeah, initially, I put all the contributors in a simple list. But interestingly, in NOMAD, all the area (D) members are equal contributors. It is nice, no! That's the reason I put all the members from area B as equal contributors. Basically, almost all of us joined in the discussion for defining grammar for NYAML.

[Nomad-joss-paper!](https://joss.theoj.org/papers/10.21105/joss.05388). But if all of you (@mkuehbach and @sherjeelshabih) want, I can remove the equal-contributor term.

    equal-contrib: true
  - given-names: Markus
    surname: Kühbach
    orcid: https://orcid.org/0000-0002-7117-5196
    affiliation: 1
    equal-contrib: true
  - given-names: Andrea
    surname: Albino
    orcid: https://orcid.org/0000-0001-9280-7431
    affiliation: 1
    equal-contrib: true
  - given-names: Florian
    surname: Dobener
    orcid: https://orcid.org/0000-0003-1987-6224
    affiliation: 1
    equal-contrib: true
  - given-names: Sherjeel
    surname: Shabih
    orcid: https://orcid.org/0009-0008-6635-4465
    affiliation: 1
    equal-contrib: true
  - given-names: Christoph
    surname: Koch
    orcid: https://orcid.org/0000-0002-3984-1523
    affiliation: 1
  - given-names: Heiko B.
    surname: Weber
    orcid: https://orcid.org/0000-0002-6403-9022
    affiliation: 2
  - given-names: José Antonio
    surname: Márquez Prieto
    orcid: https://orcid.org/0000-0002-8173-2566
    affiliation: 1
  - given-names: Claudia
    surname: Draxl
    orcid: https://orcid.org/0000-0003-3523-6657
    affiliation: 1
  - given-names: Sandor
    surname: Brockhauser
    orcid: https://orcid.org/0000-0002-9700-4803
    affiliation: 1
    equal-contrib: true

affiliations:

>   - name: Physics Department and CSMB, Humboldt-Universität zu Berlin, Berlin, Germany

Reviewer: Should have full addresses, see pynxtools draft

# Statement of need
The growth of both the standard and the number of NeXus definition developers makes it all the more important to ensure that the development process is both user-friendly and resilient to errors. The existing representation of NeXus definitions in XML offers structural rigor through the NeXus Definition Language (NXDL) and a well-defined hierarchy for metadata and data types. However, it is verbose and can be difficult to edit by hand. Writing and maintaining NXDL files often involves dealing with deeply nested elements and strict syntax, which can be error-prone and time-consuming, especially for users who are not familiar with XML development. __nyaml__ addresses these challenges by enabling the development of NeXus definitions in a cleaner, YAML-based format while preserving the full structure, semantics, and developer comments of the original XML.

# nyaml Converter

> The __nyaml__ tool is a Python package, containing the several modules, that provides a command line interface for converting NeXus application definitions or base classes from YAML format (file with a __.yaml__ extension) into the XML format (file with a __.nxdl.xml__ extension) and vice-versa. To write a NeXus application defintion or a base class in YAML format, the __nyaml__ package introduces a set of the keywords and syntactic rules (see [Documentation](https://fairmat-nfdi.github.io/nyaml)) that are specific to the NXDL (NeXus Definition Language) in the YAML format.

Reviewer: The first sentence is just a repetition of what is said in line 79

Author: Refactored!

>  The __nyaml__ package is designed to be used as a command line tool, but it can also be utilised as a Python module for programmatic use. The package in published in PyPI and therefore can be installed using python package manager __pip__. The converter command is invoked by the __nyaml2nxdl__ registered in the __nyaml.cli__ module [@Click:2025]. In that module, the function __launch_tool__ decides upon the input file type which conversion (either from YAML to XML or XML to YAML) will be invoked and execute the corresponding pipline of the data workflow (\autoref{fig:nyaml_workflow}).

Reviewer: ```suggestion
 The __nyaml__ package is designed to be used as a command line tool, but it also contains Python module for programmatic use. The package in published in PyPI and therefore can be installed using python package managers (e.g.  __pip__. ` The converter command is invoked by the __nyaml2nxdl__ registered in the __nyaml.cli__ module [@Click:2025]. In that module, the function __launch_tool__ decides upon the input file type which conversion (either from YAML to XML or XML to YAML) will be invoked and execute the corresponding pipline of the data workflow (\autoref{fig:nyaml_workflow}).
```

Author: Addressed!

    surname: Shabih
    orcid: https://orcid.org/0009-0008-6635-4465
    affiliation: 1
  - given-names: Christoph
    surname: Koch
    orcid: https://orcid.org/0000-0002-3984-1523
    affiliation: 1
  - given-names: Heiko B.
    surname: Weber
    orcid: https://orcid.org/0000-0002-6403-9022
    affiliation: 2
  - given-names: José Antonio
    surname: Márquez Prieto
    orcid: https://orcid.org/0000-0002-8173-2566
    affiliation: 1
  - given-names: Claudia
    surname: Draxl
    orcid: https://orcid.org/0000-0003-3523-6657
    affiliation: 1
  - given-names: Sandor
    surname: Brockhauser
    orcid: https://orcid.org/0000-0002-9700-4803
    affiliation: 1

affiliations:
  - name: Physics Department and CSMB, Humboldt-Universität zu Berlin, Zum Großen Windkanal 2, D-12489 Berlin, Germany
    index: 1
  - name: Lehrstuhl für Angewandte Physik, Friedrich-Alexander-Universität Erlangen-Nürnberg, Staudtstra{\ss}e 7, D-91058 Erlangen, Germany
    index: 2
  - name: Department Heterogeneous Reactions, Max Planck Institute for Chemical Energy Conversion, Stiftstraße 34-36, D-45470 Mülheim an der Ruhr, Germany
    index: 3

date: 01 July 2025
bibliography: paper.bib
---
# Summary

> The NeXus scientific data format standard [@Könnecke:2015]  [@Könnecke2006] [@Klosowski1997], which was originally introduced for neutron, X-ray, and muon science, has in recent years seen a significant enhancement across diverse scientific domains such as materials science. NeXus definitions specify the hierarchical structure and semantics of valid NeXus files and are written in XML [@xml10] using the NeXus Definition Language (NXDL), which itself is specified using XSD (XML Schema Definition).

Reviewer: Have a look at how to use multiple references: https://joss.readthedocs.io/en/latest/example_paper.html

They should be separated by `;`

  - given-names: Markus
    surname: Kühbach
    orcid: https://orcid.org/0000-0002-7117-5196
    affiliation: 1
  - given-names: Andrea
    surname: Albino
    orcid: https://orcid.org/0000-0001-9280-7431
    affiliation: 1
  - given-names: Florian
    surname: Dobener
    orcid: https://orcid.org/0000-0003-1987-6224
    affiliation: 1
  - given-names: Sherjeel
    surname: Shabih
    orcid: https://orcid.org/0009-0008-6635-4465
    affiliation: 1
  - given-names: Christoph
    surname: Koch
    orcid: https://orcid.org/0000-0002-3984-1523
    affiliation: 1
  - given-names: Heiko B.
    surname: Weber
    orcid: https://orcid.org/0000-0002-6403-9022
    affiliation: 2
  - given-names: José Antonio
    surname: Márquez Prieto
    orcid: https://orcid.org/0000-0002-8173-2566
    affiliation: 1
  - given-names: Claudia
    surname: Draxl
    orcid: https://orcid.org/0000-0003-3523-6657
    affiliation: 1
  - given-names: Sandor
    surname: Brockhauser
    orcid: https://orcid.org/0000-0002-9700-4803
    affiliation: 1

affiliations:
  - name: Physics Department and CSMB, Humboldt-Universität zu Berlin, Zum Großen Windkanal 2, D-12489 Berlin, Germany
    index: 1

>   - name: Lehrstuhl für Angewandte Physik, Friedrich-Alexander-Universität Erlangen-Nürnberg, Staudtstr. 7, D-91058 Erlangen, Germany

Reviewer 2: Staudtstra{\ss}e

Author: This is how the addresses are given in institution's website and
```
Lehrstuhl für Angewandte Physik
Friedrich-Alexander-Universität Erlangen-Nürnberg
Staudtstr. 7 / Bau A3
91058 Erlangen
```
Also, using the same address in `pynxtools` paper.

Reviewer 2: Correct and also there I kindly asked Sherjeel to write the address out like I ask here.
We should not type and discuss more than it takes us time to fix it.

    index: 2
  - name: Department Heterogeneous Reactions, Max Planck Institute for Chemical Energy Conversion, Stiftstraße 34-36, D-45470 Mülheim an der Ruhr, Germany
    index: 3

date: 01 July 2025
bibliography: paper.bib
---

# Summary

> The NeXus scientific data format standard [@Koennecke:2015], which was originally introduced for neutron, X-ray, and muon science, has in recent years seen a significant enhancement across diverse scientific domains such as materials science. NeXus definitions specify the hierarchical structure and semantics of valid NeXus files and are written in XML [@xml10] using the NeXus Definition Language (NXDL), which itself is specified using XSD (XML Schema Definition).

Reviewer 2: Könnecke2015,Könnecke2006,Klosowski1997

# Statement of need
The growth of both the standard and the number of NeXus definition developers makes it all the more important to ensure that the development process is both user-friendly and resilient to errors. The existing representation of NeXus definitions in XML offers structural rigor through the NeXus Definition Language (NXDL) and a well-defined hierarchy for metadata and data types. However, it is verbose and can be difficult to edit by hand. Writing and maintaining NXDL files often involves dealing with deeply nested elements and strict syntax, which can be error-prone and time-consuming, especially for users who are not familiar with XML development. __nyaml__ addresses these challenges by enabling the development of NeXus definitions in a cleaner, YAML-based format while preserving the full structure, semantics, and developer comments of the original XML.

> # nyaml Converter

Reviewer 2: Conversion process

> 

Reviewer 2: First paragraph shorten, everything that was  already mentioned at this point remove

Reviewer: changed

> The __nyaml__ tool is a Python package developed for converting NeXus application definitions or base classes from YAML format (files with a __.yaml__ extension) into the XML format (files with a __.nxdl.xml__ extension), and vice versa. The package is published on PyPI and therefore can be installed using Python package managers (e.g., __pip__). To write a NeXus application definition or base class in YAML format, the __nyaml__ package introduces a set of keywords and syntactic rules (see [documentation](https://fairmat-nfdi.github.io/nyaml)) that are specific to NXDL (NeXus Definition Language) in YAML.

Reviewer 2: pip pypi link to add

Author: added!

> The tool is designed to be used as a command-line utility but can also be utilized as a Python module for programmatic usages. The converter command is invoked using __nyaml2nxdl__, which determines the conversion direction based on the input file (either from YAML to XML or XML to YAML) and delegates the task to the appropriate converter. The converter then executes the corresponding data workflow pipeline (see \autoref{fig:nyaml_workflow}). These workflows are designed to ensure that the conversion process is efficient, reliable, reproducible, and preserves the integrity of the original NeXus data structure and semantics.

Reviewer 2: Could start with the sentence introducing the figure, the sentence "These workflows..." is imprecise what makes it efficient? reliable? reproducible? and integrity preserving so should be removed

Author: `efficient`: SHA256 hash mechanism makes the conversion efficient, where we build neither the `yaml` object nor the `xml` tree, but produces the `xml` file.

`reliable and reproducible`: The tool can create the same `xml` file if nothing is changed in `yaml` and does not lose any comments.

`integrity`: One can easily extend the existing nexus schemas with the help of the tools that integrates the new physical concept with the existing concepts.

Author: Added the name of the workflows to clarify "These workflows..."

Conversion from __YAML to XML__ follows specific workflow steps (depicted in \autoref{fig:nyaml_workflow}) according to NXDL rules and syntax specific to the YAML format (see [documentation](https://fairmat-nfdi.github.io/nyaml)). Starting from a given input YAML file (see \autoref{fig:nyaml_workflow}), the workflow performs the following steps:

1. The __nyaml2nxdl__ converter collects the input __.yaml__ file.

2. Using __PyYAML__ [@PyYAML:2024], the converter collects and tracks comments in the YAML file.

3. The converter parses the YAML file into a nested hashed map—a Python dictionary object.

4. The converter writes the hashed map and comments into an output XML file in accordance with NXDL concepts.

The conversion algorithm interprets the specific keywords and syntactic rules to transcode the NXDL from YAML into XML (see [documentation](https://fairmat-nfdi.github.io/nyaml)). Leveraging the NXDL rules, the conversion process detects possible inconsistencies in the YAML content and raises errors or warnings if the rules are not properly followed.

> The __XML to YAML__ conversion also follows a well-defined data workflow (\autoref{fig:nyaml_workflow}) that converts a given input XML file into a YAML file. The workflow begins with the XML input and proceeds as follows:

Reviewer 2: Move the XML to yaml block first,

always one starts with xml to yaml because still xml is the preferred choice

Author: The workflow in a order, `yaml->xml` and `xml->yaml`, it is fair to keep the order consistent.
I really not not understand how the order makes a difference. we are talking about a tool that converts from yaml ->xml and xml->  yml. It does not matter who came first and who came later in description.

Reviewer: Suggestion: add a sentence describing the difference use cases

- When editing existing definitions, workflow is XML ->YAML -> XML
- When making a new definition from scratch, you _can_ go YAML -> YML directly

1. The __nxdl2nyaml__ converter takes over the __.nxdl.xml__ file.

2. Using __lxml__ [@lxml:2025], the converter parses the XML file and builds an XML tree structure.

3. Applying YAML-specific keywords and formatting rules, the converter generates a YAML file from the XML tree.

4. The converter computes a SHA256 hash of the generated YAML content.

5. The converter writes the YAML file and appends both the hash and the original XML content as comments at the end of the YAML file.

> By attaching the hash and the original XML content to the YAML output (__.yaml__ file), the tool enables lossless round-trip conversions — provided the YAML content remains unchanged. That is, if the YAML is not modified, converting back from YAML to XML, the original commented XML content will be written back to the XML file without any modification. However, if the YAML content is modified, the XML tree will be reconstructed from the YAML content and written to the XML file, which will differ from the original XML content included in the comments. This caching approach streamlines the XML → YAML → XML workflow and facilitates straightforward comparisons of XML files in version control systems such as Git.

Reviewer 2: see comment I made on the docstring branch https://github.com/FAIRmat-NFDI/nyaml/pull/33/files#r2202707647

> By attaching the hash and the original XML content to the YAML output (__.yaml__ file), the tool enables lossless round-trip conversions — provided the YAML content remains unchanged. That is, if the YAML is not modified, converting back from YAML to XML, the original commented XML content will be written back to the XML file without any modification. However, if the YAML content is modified, the XML tree will be reconstructed from the YAML content and written to the XML file, which will differ from the original XML content included in the comments. This caching approach streamlines the XML → YAML → XML workflow and facilitates straightforward comparisons of XML files in version control systems such as Git.

Reviewer 2: "That is .." sentence remove rather we should make the previous sentence stating this clearly

# Evaluation from NIAC

> The NeXus International Advisory Committee (NIAC) is the governing body responsible for overseeing the development and maintenance of the NeXus data standard. A core responsibility of the NIAC is the stewardship of the NeXus Definition Language (NXDL), the XML-based schema that defines the hierarchical structure and semantics of NeXus data files [Koennecke:2015]. As part of its mission to facilitate the standardization of NeXus definitions in NXDL, NIAC has recently reviewed and formally accepted the __nyaml__ tool. Following a successful evaluation, NIAC has approved __nyaml__ and endorsed it as the recommended tool for preparing NeXus definition proposals. In support of this decision, the official NeXus definition repository was updated to integrate __nyaml__ into its workflow through the addition of two makefile targets: 'make nyaml', which converts existing definitions from the canonical nxdl.xml format into .yaml, and 'make nxdl', which detects modified or newly added .yaml files and converts them back into valid nxdl.xml format for submission and version control. This integration ensures that contributions made in .nyaml are compatible with the existing XML-based infrastructure. The adoption of __nyaml__ by NIAC reflects an ongoing commitment to fostering community engagement and modernizing the technical tools underpinning the NeXus standard [@NIAC:2025].

Reviewer 2: "the XML schema" > "the XML schemas"

Author: addresed!

> The NeXus International Advisory Committee (NIAC) is the governing body responsible for overseeing the development and maintenance of the NeXus data standard. A core responsibility of the NIAC is the stewardship of the NeXus Definition Language (NXDL), the XML-based schema that defines the hierarchical structure and semantics of NeXus data files [Koennecke:2015]. As part of its mission to facilitate the standardization of NeXus definitions in NXDL, NIAC has recently reviewed and formally accepted the __nyaml__ tool. Following a successful evaluation, NIAC has approved __nyaml__ and endorsed it as the recommended tool for preparing NeXus definition proposals. In support of this decision, the official NeXus definition repository was updated to integrate __nyaml__ into its workflow through the addition of two makefile targets: 'make nyaml', which converts existing definitions from the canonical nxdl.xml format into .yaml, and 'make nxdl', which detects modified or newly added .yaml files and converts them back into valid nxdl.xml format for submission and version control. This integration ensures that contributions made in .nyaml are compatible with the existing XML-based infrastructure. The adoption of __nyaml__ by NIAC reflects an ongoing commitment to fostering community engagement and modernizing the technical tools underpinning the NeXus standard [@NIAC:2025].

Reviewer 2: recently, we should replace with link to the pr where this was decided

Reviewer 2: "Following a successful" this sentence is a duplicate just continue with "In support of..."

Reviewer 2: Not sure whether we should explain the addition of the two maekfile targets that long maybe just

In support of this decision [add ref to that pr or release] two makefile targets were added emdash make nyaml and make nxdl emdash for converting to YAML and XML respectively.

Reviewer 2: The last sentence is a testimonial statement that we may shorten but indeed could keep on the other hand if the standardization body wouldnt do anything it is dead

Reviewer: Regarding the first comment, I think this is the PR: https://github.com/nexusformat/definitions/pull/1336

Reviewer: Add a link to the https://github.com/nexusformat/definitions when you talk about the official definitions

# Acknowledgements

> The __nyaml__ software development is funded by the German National Research Data Infrastructure

Reviewer 2: we should check that this follows exactly the parameterized statement requested by the dfg

Author: This is the example: 
```Example: "This work was supported by the National Research Data Infrastructure (NFDI) consortium [NFDI consortium name] – [Reference number]" ``` used for it.

---
title: 'nyaml: A Format Converter for NeXus Data Model.'
tags:
    - Python
    - NeXus
    - NXDL
authors:

>   - name: Adrian M. Price-Whelan

Reviewer: Add correct authors here, see https://github.com/FAIRmat-NFDI/pynxtools/pull/658/files for most ORCIDS etc.

Reviewer 2: exactly, in addition use the 
 equal-contrib for the following authors:
 * Rubel (many implementation details)
 
 we should negotiate as to how we order these three authors
 * Andrea (handover from Markus further functionalities)
 * Lukas (many bugfixes and improved tests)
 * Markus (original implementation from scratch plus bugfixes later), 
 
 all remaining authors do not have the equal contributions

Author: It could be something like:
1. Rubel, 2. Lukas, 3. Markus 4. Andrea

Reviewer 2: @RubelMozumder please make your proposal straight via a commit, suggestions were specific, so cycle should be implement, commit, review, implement, commit, review, ..., resolve, merge

Author: All the area-B members are equal contributors.

    orcid: 0000-0000-0000-0000
    equal-contrib: true
    affiliation: "1, 2" # (Multiple affiliations must be quoted)
  - name: Author Without ORCID
    equal-contrib: true # (This is how you can denote equal contributions between multiple authors)
    affiliation: 2
  - name: Author with no affiliation
    corresponding: true # (This is how to denote the corresponding author)
    affiliation: 3
  - given-names: Ludwig
    dropping-particle: van
    surname: Beethoven
    affiliation: 3
affiliations:
 - name: Lyman Spitzer, Jr. Fellow, Princeton University, United States
   index: 1
   ror: 00hx57361
 - name: Institution Name, Country
   index: 2
 - name: Independent Researcher, Country
   index: 3

> date: 13 August 2017

Author: Update!

---

> title: 'nyaml: A Format Converter for NeXus Data Model.'

Reviewer 2: "A" not to be capitalized

add "the" after "for" i.e. "for the NeXus"

> tags:

Reviewer 2: add:
YAML
XML

Author: added!

    - Python
    - NeXus
    - NXDL
authors:
  - name: Adrian M. Price-Whelan
    orcid: 0000-0000-0000-0000
    equal-contrib: true
    affiliation: "1, 2" # (Multiple affiliations must be quoted)
  - name: Author Without ORCID
    equal-contrib: true # (This is how you can denote equal contributions between multiple authors)
    affiliation: 2
  - name: Author with no affiliation
    corresponding: true # (This is how to denote the corresponding author)
    affiliation: 3
  - given-names: Ludwig
    dropping-particle: van
    surname: Beethoven
    affiliation: 3
affiliations:
 - name: Lyman Spitzer, Jr. Fellow, Princeton University, United States
   index: 1
   ror: 00hx57361
 - name: Institution Name, Country
   index: 2
 - name: Independent Researcher, Country
   index: 3
date: 13 August 2017
bibliography: paper.bib
---
# Summary
# Statement of need
# NYAML to NXDL
# NXDL to NYAML

> # Evaluation from NAIC

Reviewer 2: Evaluation from the NIAC

Reviewer 2: We shall though not have these acronyms in the section header unless we are sure to have these introduced in preceeding sections, overall though spelling these acronyms out is better is my opinion

Author: Abbreviation of the acronyms is included.

# Citations

NOTE! The following is example of citation
Citations to entries in paper.bib should be in
[rMarkdown](http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html)
format.

If you want to cite a software repository URL (e.g. something on GitHub without a preferred
citation) then you can do it with the example BibTeX entry below for @fidgit.

For a quick reference, the following citation commands can be used:
- `@author:2001`  ->  "Author et al. (2001)"
- `[@author:2001]` -> "(Author et al., 2001)"
- `[@author1:2001; @author2:2001]` -> "(Author1 et al., 2001; Author2 et al., 2002)"

# Figures

> Note! The follwoing is example figure

Reviewer 2: see comment for citation section above

Figures can be included like this:
![Caption for example figure.\label{fig:example}](figure.png)
and referenced from text using \autoref{fig:example}.

Figure sizes can be customized by adding an optional second parameter:
![Caption for example figure.](figure.png){ width=20% }

> # Acknowledgements

Reviewer 2: add # Author contributions

Author: Mainly focused on the funding agency (FAIRmat), as Nomad and other FAIRmat papers are acknowledge.

    index: 3

date: 01 July 2025
bibliography: paper.bib
---

# Summary

The NeXus scientific data format standard [@Könnecke:2015; @Könnecke2006; @Klosowski1997], which was originally introduced for neutron, X-ray, and muon science, has in recent years seen a significant enhancement across diverse scientific domains such as materials science. NeXus definitions—comprising application definitions and base classes—describe the hierarchical structure and semantics of valid NeXus files. They are written in XML [@xml10] using the NeXus Definition Language (NXDL), which is itself defined in XSD (XML Schema Definition) [@w3c:xmlschema1].

`nyaml` is a Python-based tool with both a command-line interface and an application programming interface (API) that facilitates the conversion between NXDL XML and a simplified YAML [@YAML11] representation. YAML's indentation-based syntax enhances human readability and simplifies manual editing. By providing a reliable, lossless round-trip conversion between XML and YAML, `nyaml` enables developers to edit NeXus definitions efficiently without sacrificing structural or semantic fidelity.

# Statement of need

> As the NeXus standard and its community of definition developers continue to grow, it is increasingly important to ensure that the development process remains both user-friendly and resilient to errors. The existing representation of NeXus definitions in XML offers structural rigor through the NeXus Definition Language (NXDL) and a well-defined hierarchy for metadata and data types. However, it is verbose and can be difficult to edit by hand. Writing and maintaining NXDL files often involves dealing with deeply nested elements and strict syntax, which can be error-prone and time-consuming, especially for users who are not familiar with XML development. `nyaml` addresses these challenges by enabling the development of NeXus definitions in a simpler, YAML-based format while preserving the full structure, semantics, and developer comments of the original XML. As a use case, the tool has been used to integrate NeXus definitions into the large-scale research data management platform NOMAD [@NOMAD2023], with bidirectional YAML–XML conversions enabling modifications and extensions to make the NeXus schemata compatibility with the NOMAD data model and other physics communities.

Reviewer 2: "As a use case, the tool has been used to integrate NeXus definitions into the large-scale research data management platform NOMAD [@NOMAD2023], with bidirectional YAML–XML conversions enabling modifications and extensions to make the NeXus schemata compatibility with the NOMAD data model and other physics communities."

reformulate like so:
"As a successful stress test, the tool was used for developing extensions of the NeXus definitions for different microscopy and spectroscopy methods to integrate these into the research data management system NOMAD [@NOMAD2023]."

"was used" or "has been used" I don't mind, use what fits best with the rest of the paragraph

Reviewer 2: Given that we already use references in the summary we could also move this statement on NOMAD in between "YAML representation" and "YAML integration" into the summary/abstract. I think talking about an application in an statement of needs is not a perfect fit while making that use case an own small paragraph is too little a text to justify that paragraph.

Author: Text is updated as you suggested.

    surname: Shabih
    orcid: https://orcid.org/0009-0008-6635-4465
    affiliation: 1
  - given-names: Christoph
    surname: Koch
    orcid: https://orcid.org/0000-0002-3984-1523
    affiliation: 1
  - given-names: Heiko B.
    surname: Weber
    orcid: https://orcid.org/0000-0002-6403-9022
    affiliation: 2
  - given-names: José Antonio
    surname: Márquez Prieto
    orcid: https://orcid.org/0000-0002-8173-2566
    affiliation: 1
  - given-names: Claudia
    surname: Draxl
    orcid: https://orcid.org/0000-0003-3523-6657
    affiliation: 1
  - given-names: Sandor
    surname: Brockhauser
    orcid: https://orcid.org/0000-0002-9700-4803
    affiliation: 1

affiliations:
  - name: Physics Department and CSMB, Humboldt-Universität zu Berlin, Zum Großen Windkanal 2, D-12489 Berlin, Germany
    index: 1
  - name: Lehrstuhl für Angewandte Physik, Friedrich-Alexander-Universität Erlangen-Nürnberg, Staudtstraße 7, D-91058 Erlangen, Germany
    index: 2
  - name: Department Heterogeneous Reactions, Max Planck Institute for Chemical Energy Conversion, Stiftstraße 34-36, D-45470 Mülheim an der Ruhr, Germany
    index: 3

date: 01 July 2025
bibliography: paper.bib
---
# Summary

> The NeXus scientific data format standard [@Könnecke:2015; @Könnecke2006; @Klosowski1997], which was originally introduced for neutron, X-ray, and muon science, has in recent years seen a significant enhancement across diverse scientific domains such as materials science. NeXus definitions, consisting of application definitions or base classes, specify the hierarchical structure and semantics of valid NeXus files and are written in XML [@xml10] using the NeXus Definition Language (NXDL), which itself is specified using XSD (XML Schema Definition).

Reviewer: Maybe we can have a reference for XSD?

Author: Added!

> The NeXus scientific data format standard [@Könnecke:2015; @Könnecke2006; @Klosowski1997], which was originally introduced for neutron, X-ray, and muon science, has in recent years seen a significant enhancement across diverse scientific domains such as materials science. NeXus definitions, consisting of application definitions or base classes, specify the hierarchical structure and semantics of valid NeXus files and are written in XML [@xml10] using the NeXus Definition Language (NXDL), which itself is specified using XSD (XML Schema Definition) [@w3c:xmlschema1].

Reviewer: add "xmlschema" to custom-dictionary

> __nyaml__ is a Python-based tool with both a command-line interface and an importable API that facilitates the conversion between NXDL XML and a simplified YAML [@YAML11] representation. YAML's indentation-based syntax enhances human readability and simplifies manual editing. By providing a reliable, lossless round-trip conversion between XML and YAML, __nyaml__ enables developers to edit NeXus definitions more efficiently without sacrificing structural or semantic fidelity.

Reviewer: I suggest we change naming of software tools using ``, like in the [JOSS example paper](https://joss.readthedocs.io/en/latest/example_paper.html). So, for this part:
```suggestion
`nyaml` is a Python-based tool with both a command-line interface and an importable API that facilitates the conversion between NXDL XML and a simplified YAML [@YAML11] representation. YAML's indentation-based syntax enhances human readability and simplifies manual editing. By providing a reliable, lossless round-trip conversion between XML and YAML, `nyaml` enables developers to edit NeXus definitions more efficiently without sacrificing structural or semantic fidelity.
```

and same below

# Statement of need

> The growth of both the standard and the number of NeXus definition developers makes it all the more important to ensure that the development process is both user-friendly and resilient to errors. The existing representation of NeXus definitions in XML offers structural rigor through the NeXus Definition Language (NXDL) and a well-defined hierarchy for metadata and data types. However, it is verbose and can be difficult to edit by hand. Writing and maintaining NXDL files often involves dealing with deeply nested elements and strict syntax, which can be error-prone and time-consuming, especially for users who are not familiar with XML development. __nyaml__ addresses these challenges by enabling the development of NeXus definitions in a cleaner, YAML-based format while preserving the full structure, semantics, and developer comments of the original XML.

Reviewer: ```suggestion
The growth of both the standard and the number of NeXus definition developers makes it all the more important to ensure that the development process is both user-friendly and resilient to errors. The existing representation of NeXus definitions in XML offers structural rigor through the NeXus Definition Language (NXDL) and a well-defined hierarchy for metadata and data types. However, it is verbose and can be difficult to edit by hand. Writing and maintaining NXDL files often involves dealing with deeply nested elements and strict syntax, which can be error-prone and time-consuming, especially for users who are not familiar with XML development. `nyaml` addresses these challenges by enabling the development of NeXus definitions in a simpler, YAML-based format while preserving the full structure, semantics, and developer comments of the original XML.
```

# Conversion process

> The __nyaml__ tool is a Python package developed for converting NeXus definitions from YAML format (files with a __.yaml__ extension) into the XML format (files with a __.nxdl.xml__ extension), and vice versa. The package is published on PyPI and therefore can be installed using Python package managers (e.g., __pip__ [@pip2025]). To write or modify NeXus definitions in YAML format, modification of existing definitions the conversion goes through __XML $\rightarrow$ YAML $\rightarrow$ modification of YAML $\rightarrow$ XML__ and new definitions that go through __YAML $\rightarrow$ XML__, the __nyaml__ package introduces a set of keywords and syntactic rules (see [documentation](https://fairmat-nfdi.github.io/nyaml)) that are specific to NXDL (NeXus Definition Language) in YAML.

Reviewer: ```suggestion
The `nyaml` tool is a Python package developed for converting NeXus definitions from YAML format (files with a `.yaml` extension) into the XML format (files with a `.nxdl.xml` extension), and vice versa. The package is published on PyPI and therefore can be installed using Python package managers (e.g., `pip` [@pip2025]). The `nyaml` package introduces a set of keywords and syntactic rules (see [documentation](https://fairmat-nfdi.github.io/nyaml)) that are specific to NXDL (NeXus Definition Language) in YAML. Existing definitions can be modified using the following workflow: __XML $\rightarrow$ YAML $\rightarrow$ modification of YAML $\rightarrow$ XML__. New definitions can be designed in YAML and converted to XML directly 
```

> The tool is designed to be used as a command-line utility but can also be utilized as a Python module for programmatic usages. The converter command is invoked using __nyaml2nxdl__, which determines the conversion direction based on the input file (either from YAML to XML or XML to YAML) and delegates the task to the appropriate converter. The converter then executes the corresponding data workflow pipeline (see \autoref{fig:nyaml_workflow}). These workflows, __YAML to XML__ and __XML to YAML__, are designed to ensure that the conversion process is efficient, reliable, reproducible, and preserves the integrity of the original NeXus data structure and semantics.

Reviewer: ```suggestion
The tool is designed to be used as a command-line utility, but can also be utilized as a Python module for programmatic usages. The converter command is invoked using `nyaml2nxdl`, which determines the conversion direction based on the input file (either from YAML to XML or XML to YAML) and delegates the task to the appropriate converter. The converter then executes the corresponding data workflow pipeline (see \autoref{fig:nyaml_workflow}). These workflows, YAML to XML and XML to YAML. are designed to ensure that the conversion process is efficient, reliable, reproducible, and preserves the integrity of the original NeXus data structure and semantics.
```

![Converter workflow for nyaml tool, XML to YAML and vice-versa \label{fig:nyaml_workflow}](assets/workflow-1.pdf){ width=75% }

> Conversion from __YAML to XML__ follows specific workflow steps (depicted in \autoref{fig:nyaml_workflow}). Given an input YAML file (see \autoref{fig:nyaml_workflow}), the __nyaml2nxdl__ converter reads the __.yaml__ file, collects and tracks the locations of comments, and then constructs a nested Python hash-mapped object representing the YAML content using __PyYAML__ [@PyYAML:2024]. In the final stage, in accordance with the NXDL grammar and syntax specific to the XML format, the converter generates an XML file that includes the collected comments. In the __YAML to XML__ conversion, the algorithm interprets the keywords and syntactic rules specific to the YAML format (see [documentation](https://fairmat-nfdi.github.io/nyaml)) to transcode the NXDL content from YAML into XML. Leveraging the NXDL rules, the conversion process detects possible inconsistencies in the YAML content and raises errors or warnings if the rules are not properly followed.

Reviewer: ```suggestion
Conversion from __YAML to XML__ follows specific workflow steps (depicted in \autoref{fig:nyaml_workflow}). Given an input YAML file (see \autoref{fig:nyaml_workflow}), the `nyaml2nxdl` converter reads the `.yaml` file, collects and tracks the locations of comments, and then constructs a nested Python hash-mapped object representing the YAML content using `PyYAML` [@PyYAML:2024]. In the final stage, in accordance with the NXDL grammar and syntax specific to the XML format, the converter generates an XML file that includes the collected comments. In the __YAML to XML__ conversion, the algorithm interprets the keywords and syntactic rules specific to the YAML format (see [documentation](https://fairmat-nfdi.github.io/nyaml)) to transcode the NXDL content from YAML into XML. Leveraging the NXDL rules, the conversion process detects possible inconsistencies in the YAML content and raises errors or warnings if the rules are not properly followed.
```

> In the reverse direction — __XML to YAML__ conversion (see \autoref{fig:nyaml_workflow}) — the __nxdl2nyaml__ converter takes the input __.nxdl.xml__ file and transforms it into an XML tree structure using the __lxml__ library [@lxml:2025]. This XML tree is then transcoded into a YAML file that adheres to the NXDL rules and syntax specific to the YAML format. The converter then computes a SHA256 hash from the generated YAML content and appends the hash to the end of the YAML file, followed by the original XML content. By attaching both the hash and the original XML content to the YAML output (__.yaml__ file), the tool enables lossless round-trip conversions: if the YAML is not modified, then when converting back from YAML to XML, the original commented XML content will be restored in the output XML file without modification. However, if the YAML content is changed, the XML tree will be reconstructed from the modified YAML and written to the XML file, resulting in differences from the original XML content included in the comments. This caching approach streamlines the XML → YAML → XML workflow and facilitates straightforward comparisons of XML files in version control systems such as Git.

Reviewer: ```suggestion
In the reverse direction — __XML to YAML__ conversion (see \autoref{fig:nyaml_workflow}) — the `nxdl2nyaml` converter takes the input `.nxdl.xml` file and transforms it into an XML tree structure using the `lxml` library [@lxml:2025]. This XML tree is then transcoded into a YAML file that adheres to the NXDL rules and syntax specific to the YAML format. The converter then computes a SHA256 hash from the generated YAML content and appends the hash to the end of the YAML file, followed by the original XML content. By attaching both the hash and the original XML content to the YAML output (`.yaml` file), the tool enables lossless round-trip conversions: if the YAML is not modified, then  the original commented XML content will be restored in the output XML file without modification when converting from YAML to XML. However, if the YAML content is changed, the XML tree will be reconstructed from the modified YAML and written to the XML file, resulting in differences from the original XML content included as comments. This caching approach streamlines the XML → YAML → XML workflow and facilitates straightforward comparisons of XML files in version control systems such as Git.
```

# Evaluation from NIAC

> The NeXus International Advisory Committee (NIAC) is the governing body responsible for overseeing the development and maintenance of the NeXus data standard. A core responsibility of the NIAC is the stewardship of the NeXus Definition Language (NXDL), the XML-based definitions that define the hierarchical structure and semantics of NeXus data files [Könnecke:2015]. As part of its mission to facilitate the standardization of NeXus definitions in NXDL, NIAC has recently reviewed and formally accepted the __nyaml__ tool. Following a successful evaluation, NIAC has approved __nyaml__ and endorsed it as the recommended tool for preparing NeXus definition proposals. In support of this decision, the official NeXus definition repository [@nexusDef:2024] was updated to integrate __nyaml__ into its workflow through the addition of two makefile targets: 'make nyaml' [@nyamlIntegration:2024], which converts existing definitions from the canonical nxdl.xml format into .yaml, and 'make nxdl', which detects modified or newly added .yaml files and converts them back into valid nxdl.xml format for submission and version control. This integration ensures that contributions made in .nyaml are compatible with the existing XML-based infrastructure. The adoption of __nyaml__ by NIAC reflects an ongoing commitment to fostering community engagement and modernizing the technical tools underpinning the NeXus standard [@NIAC:2025].

Reviewer: ```suggestion
The NeXus International Advisory Committee (NIAC) is the governing body responsible for overseeing the development and maintenance of the NeXus data standard. A core responsibility of the NIAC is the stewardship of the NeXus Definition Language (NXDL), the XML-based definitions that define the hierarchical structure and semantics of NeXus data files [Könnecke:2015]. As part of its mission to facilitate the standardization of NeXus definitions in NXDL, NIAC has recently reviewed and formally accepted the `nyaml` tool. Following a successful evaluation, NIAC has approved `nyaml` and endorsed it as the recommended tool for preparing NeXus definition proposals. In support of this decision, the official NeXus definition repository [@nexusDef:2024] was updated to integrate `nyaml` into its workflow through the addition of two makefile targets: 'make nyaml' [@nyamlIntegration:2024], which converts existing definitions from the canonical nxdl.xml format into .yaml, and 'make nxdl', which detects modified or newly added .yaml files and converts them back into valid nxdl.xml format for submission and version control. This integration ensures that contributions made in .nyaml are compatible with the existing XML-based infrastructure. The adoption of `nyaml` by NIAC reflects an ongoing commitment to fostering community engagement and modernizing the technical tools underpinning the NeXus standard [@NIAC:2025].
```

# Acknowledgements

> The __nyaml__ software development is funded by the German National Research Data Infrastructure

Reviewer: ```suggestion
The `nyaml` software is developed by FAIRmat. FAIRmat is funded by the Deutsche Forschungsgemeinschaft (DFG, German Research Foundation) – project 460197019.
```
